### DISCUSSION TITLE
Admin boundary trouble

### DISCUSSION THREAD

**Discussion by jeffricks** [onthegomap/planetiler#850] [Q&A]
Hi,

I'm working with OSM data for India.  To get the proper admin boundaries for India, I have performed the following steps.  The issue I'm having is that when I create vector tiles from the data, some unexpected data seems to be included and some expected data gets lost.  While I'm dealing with just the boundary data below, my end goal is to somehow "augment" the boundaries from the original country-specific extract (or planet
) with the official admin boundaries.

1. Ran this Overpass query in Overpass Turbo and saved as india-admin-boundaries.osm:

```
[timeout:125];
(
  // query part for: “admin_level=2 and "ISO3166-1"=IN”
  relation["admin_level"="2"]["ISO3166-1"="IN"]({{bbox}});
  relation["claimed_by"="IN"]({{bbox}});
);

out body;
>;
out skel qt;
```

2. Used osmium to sort the data to prepare for use with planetiler:

`osmium sort -v -o /dataout/india-admin-boundaries-sorted.osm /dataout/india-admin-boundaries.osm`

2. Viewed the resulting file india-admin-boundaries-sorted.osm in Merkaator and observed that the boundaries look as expected

<img width="929" alt="2024-03-21 10_54_44-untitled - Merkaartor v0 19 0" src="https://github.com/onthegomap/planetiler/assets/3695125/45b0599d-4102-4dde-bf6f-17f50844f2ea">

3. Used osmium to convert india-admin-boundaries-sorted.osm to india-admin-boundaries-sorted.osm.pbf:

`osmium cat -v -o /dataout/india-admin-boundaries-sorted.osm.pbf /dataout/india-admin-boundaries-sorted.osm`

The resulting .osm.pbf file can be loaded with QGIS using the QuickOSM plugin successfully and it reflects what was seen in Merkaator.

<img width="816" alt="2024-03-21 20_08_03-_Untitled Project — QGIS" src="https://github.com/onthegomap/planetiler/assets/3695125/56485354-06e7-41ed-b007-747da67140b3">

4. Used planetiler to create vector tiles in MBTiles format:

`docker run --rm --name planetiler -e JAVA_TOOL_OPTIONS="-Xmx16g" -v /dataout:/data ghcr.io/onthegomap/planetiler:latest --osm-path=/dataout/india-admin-boundaries-sorted.osm.pbf --output=/dataout/india-admin-boundaries-sorted.osm.pbf.mbtiles --force`

5. Planetiler completes with no visible issues and produces the MBTiles file.

6. Hosted the MBTiles file using tileserver-gl

7. Viewed the data via three means:
	a. In QGIS - At zoom levels below 6, I see this.  I assume that geometry is coming from NE:

<img width="910" alt="2024-03-21 20_08_58-_Untitled Project — QGIS" src="https://github.com/onthegomap/planetiler/assets/3695125/0d8db3cb-c930-458c-9052-b88226f04f4f">

       At zoom levels 6+, I see this which is the geometry from my .osm.pbf file MINUS the "disputed regions" (left: generated MBTiles, right: .osm.pbf file):

<img width="1703" alt="2024-03-21 20_13_48-2024-03-21 20_08_03-_Untitled Project — QGIS png ‎- Photos" src="https://github.com/onthegomap/planetiler/assets/3695125/0bc38261-3ba2-4360-ba6d-926f7cf6c288">

        b. On the tileserver-gl "home page", clicked the "Inspect" button next to the MBTiles file and viewed the data on the displayed map - same observations.

	c. Pointed Maputnik to the hosted endpoint and viewed the data via both the Map (with my style) and Inspect views - same observations.

Screenshot showing Merkaator on left, Maputnik on right (red oval below shows the northern area where Merkaator shows more data):

<img width="1795" alt="2024-03-21 12_20_07-Merkaator-Maputnik " src="https://github.com/onthegomap/planetiler/assets/3695125/872f614a-2a3b-43f4-a65d-34c0e65eb546">

I feel like I'm just missing some required piece of the process.  I saw somewhere that Natural Earth admin 0 boundaries are used by planetiler for low zoom levels.  I tried to use the `--boundary-osm-only=true` flag but I still got the admin 0 geometry.  I also tried to modify the Natural Earth sqlite DB to make the `ne_10m_admin_0_countries` table contain the data from the `ne_10m_admin_0_countries_ind` table but that also seemed to not have an effect.

Any help or suggestions would be greatly appreciated.

Thanks,

Jeff

**Comment by phanecak-maptiler** [comment]
Not a full answer but at least a hint and pointing some direction:
- You're using OpenMapTiles profile, hence this code is generating the `boundary` layer: https://github.com/openmaptiles/planetiler-openmaptiles/blob/main/src/main/java/org/openmaptiles/layers/Boundary.java
- We can see there that (roughly):
  - NE data is being used for Z0-Z4
  - OSM data is being used for Z4+
- Given how the `minzoom` is calculated for OSM data, disputed borders show up on Z12+ since they tend to not have `admin_level` attribute
  - see https://github.com/openmaptiles/planetiler-openmaptiles/blob/main/src/main/java/org/openmaptiles/layers/Boundary.java#L266
- `--boundary-osm-only=true` should work (have not tried myself) but again if a border does not have `admin_level=2` it will not be included between Z0 and Z4
  - see https://github.com/openmaptiles/planetiler-openmaptiles/blob/main/src/main/java/org/openmaptiles/layers/Boundary.java#L271

So, how to get out of this: You can create your own "profile" and implement the logic as you need:
- In the Java: see for example https://github.com/onthegomap/planetiler/blob/main/planetiler-examples or https://github.com/openmaptiles/planetiler-openmaptiles/tree/main?tab=readme-ov-file#customizing
- Using YAML: see https://github.com/onthegomap/planetiler/blob/main/planetiler-custommap

**Reply by jeffricks** [reply]
Thanks, that helps.  I'll have to experiment some more to better understand why the `--boundary-osm-only=true` option didn't exclude the NE data.  I'll give the custom yaml config a shot... is there way to only override the parts you want to affect?  If not, is there a yaml file that contains all of planetiler's default settings that I could tweak to fix my issue?  I'm afraid that trying to reproduce all of the defaults myself could be challenging.

The other thought I had was to modify the NE sqlite data as happens for the monaco example.  I did try that but I modified the `ne_10m_admin_0_countries` table instead of `ne_10m_admin_0_boundary_lines_land` table so it didn't have any effect based on what it looks like the code is doing.

**Reply by jeffricks** [reply]
***UPDATE***: I believe the `--boundary-osm-only=true` flag IS actually working.  Now my tiles have the input MINUS the same missing areas so I think I'm running into either a `minzoom` issue or the data not having `admin_level` set.  The missing data doesn't appear at any zoom levels so it seems like it's being filtered out somehow.

**Reply by msbarry** [reply]
Thanks for helping investigate @phanecak-maptiler! @jeffricks you're trying to end up with a global openmaptiles build with India borders reflecting different logic than default profile right? If so you might want to add some specific logic to boundary layer that sets an admin level or min zoom when claimed_by=IN and use `--boundary-osm-only=true` to force it to show up at lower zoom levels?

I find it easiest to debug these issues by adding a test case to BoundaryTest.java with the specific relation/way tags that the real features have then tweaking the logic until tags/minzoom match what you want.

**Reply by jeffricks** [reply]
@msbarry, thanks for the reply.  My apologies for the delay in responding.  I finally have planetiler set up for development and I've been able to successfully run all the tests.  Can you give me an example of how I would add a test case as you've described?  I see the various test cases in BoundaryTest.java but I'm not sure how to test if my data is correct or even what the test would tell me.  Any pointers would be appreciated.  Thanks, Jeff.
